**Resource:** https://github.com/aws-samples/Meta-Llama-on-AWS/blob/main/RAG-recipes/llama3-rag-langchain-smjs.ipynb

* Embedding Model Used: *HuggingFace BGE Large EN Embedding model*
* LLM Used: *HuggingFace Llama 3 8b Instruct LLM model*

In [1]:
%%writefile requirements.txt
langchain==0.1.14
pypdf==4.1.0
faiss-cpu==1.8.0
boto3==1.34.58
sqlalchemy==2.0.29

Overwriting requirements.txt


In [2]:
import sqlalchemy
print(sqlalchemy.__version__)

2.0.30


In [3]:
pip install nvidia-ml-py3==7.352.0

  Using cached nvidia_ml_py3-7.352.0-py3-none-any.whl
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
autogluon-multimodal 1.1.1 requires omegaconf<2.3.0,>=2.1.1, but you have omegaconf 2.3.0 which is incompatible.
autogluon-multimodal 1.1.1 requires scikit-learn<1.4.1,>=1.3.0, but you have scikit-learn 1.4.2 which is incompatible.
Note: you may need to restart the kernel to use updated packages.


In [4]:
pip install sqlparse==0.5.0

  Using cached sqlparse-0.5.0-py3-none-any.whl.metadata (3.9 kB)
Using cached sqlparse-0.5.0-py3-none-any.whl (43 kB)
  Attempting uninstall: sqlparse
    Found existing installation: sqlparse 0.5.1
    Uninstalling sqlparse-0.5.1:
      Successfully uninstalled sqlparse-0.5.1
Note: you may need to restart the kernel to use updated packages.


In [5]:
pip install scikit-learn==1.3.0

  Using cached scikit_learn-1.3.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (11 kB)
Using cached scikit_learn-1.3.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (10.9 MB)
  Attempting uninstall: scikit-learn
    Found existing installation: scikit-learn 1.4.2
    Uninstalling scikit-learn-1.4.2:
      Successfully uninstalled scikit-learn-1.4.2
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
autogluon-multimodal 1.1.1 requires omegaconf<2.3.0,>=2.1.1, but you have omegaconf 2.3.0 which is incompatible.
autogluon-timeseries 1.1.1 requires gluonts==0.15.1, but you have gluonts 0.14.3 which is incompatible.
Note: you may need to restart the kernel to use updated packages.


In [6]:
pip install omegaconf==2.2.3

  Using cached omegaconf-2.2.3-py3-none-any.whl.metadata (3.9 kB)
Using cached omegaconf-2.2.3-py3-none-any.whl (79 kB)
  Attempting uninstall: omegaconf
    Found existing installation: omegaconf 2.3.0
    Uninstalling omegaconf-2.3.0:
      Successfully uninstalled omegaconf-2.3.0
Note: you may need to restart the kernel to use updated packages.


In [7]:
pip install gluonts==0.15.1

  Using cached gluonts-0.15.1-py3-none-any.whl.metadata (9.9 kB)
Using cached gluonts-0.15.1-py3-none-any.whl (1.5 MB)
  Attempting uninstall: gluonts
    Found existing installation: gluonts 0.14.3
    Uninstalling gluonts-0.14.3:
      Successfully uninstalled gluonts-0.14.3
Note: you may need to restart the kernel to use updated packages.


In [8]:
pip install langchain==0.1.14

  Using cached langchain-0.1.14-py3-none-any.whl.metadata (13 kB)
  Using cached langchain_community-0.0.38-py3-none-any.whl.metadata (8.7 kB)
  Using cached langchain_core-0.1.52-py3-none-any.whl.metadata (5.9 kB)
  Using cached langchain_text_splitters-0.0.2-py3-none-any.whl.metadata (2.2 kB)
  Using cached packaging-23.2-py3-none-any.whl.metadata (3.2 kB)
Using cached langchain-0.1.14-py3-none-any.whl (812 kB)
Using cached langchain_community-0.0.38-py3-none-any.whl (2.0 MB)
Using cached langchain_core-0.1.52-py3-none-any.whl (302 kB)
Using cached langchain_text_splitters-0.0.2-py3-none-any.whl (23 kB)
Using cached packaging-23.2-py3-none-any.whl (53 kB)
  Attempting uninstall: packaging
    Found existing installation: packaging 24.1
    Uninstalling packaging-24.1:
      Successfully uninstalled packaging-24.1
  Attempting uninstall: langchain-core
    Found existing installation: langchain-core 0.2.33
    Uninstalling langchain-core-0.2.33:
      Successfully uninstalled langchai

In [9]:
pip install boto3==1.34.58

  Using cached boto3-1.34.58-py3-none-any.whl.metadata (6.6 kB)
Using cached boto3-1.34.58-py3-none-any.whl (139 kB)
  Attempting uninstall: boto3
    Found existing installation: boto3 1.34.131
    Uninstalling boto3-1.34.131:
      Successfully uninstalled boto3-1.34.131
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
langchain-aws 0.1.16 requires boto3<1.35.0,>=1.34.131, but you have boto3 1.34.58 which is incompatible.
langchain-aws 0.1.16 requires langchain-core<0.3,>=0.2.29, but you have langchain-core 0.1.52 which is incompatible.
sagemaker 2.227.0 requires boto3<2.0,>=1.34.142, but you have boto3 1.34.58 which is incompatible.
Note: you may need to restart the kernel to use updated packages.


In [10]:
!pip install -U -r requirements.txt

  Using cached pypdf-4.1.0-py3-none-any.whl.metadata (7.4 kB)
  Using cached faiss_cpu-1.8.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (3.6 kB)
  Using cached SQLAlchemy-2.0.29-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl.metadata (9.6 kB)
Using cached pypdf-4.1.0-py3-none-any.whl (286 kB)
Using cached faiss_cpu-1.8.0-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (27.0 MB)
Using cached SQLAlchemy-2.0.29-cp311-cp311-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (3.2 MB)
  Attempting uninstall: sqlalchemy
    Found existing installation: SQLAlchemy 2.0.30
    Uninstalling SQLAlchemy-2.0.30:
      Successfully uninstalled SQLAlchemy-2.0.30


In [11]:
import langchain
print(langchain.__version__)

0.1.14


In [12]:
try:
    import sagemaker
except ImportError:
    !pip install sagemaker

sagemaker.config INFO - Not applying SDK defaults from location: /etc/xdg/sagemaker/config.yaml
sagemaker.config INFO - Not applying SDK defaults from location: /home/sagemaker-user/.config/sagemaker/config.yaml


In [13]:
# Import the JumpStartModel class from the SageMaker JumpStart library
from sagemaker.jumpstart.model import JumpStartModel

In [14]:
# Specify the model ID for the HuggingFace Llama 3 8b Instruct LLM model
model_id = "meta-textgeneration-llama-3-8b-instruct"
accept_eula = True
model = JumpStartModel(model_id=model_id, model_version="2.7.0", instance_type= "ml.g5.2xlarge")

In [15]:
predictor = model.deploy(accept_eula=accept_eula, instance_type="ml.g5.2xlarge")

------------------!

In [16]:
# Specify the model ID for the HuggingFace BGE Large EN Embedding model
model_id = "huggingface-sentencesimilarity-bge-large-en-v1-5"
text_embedding_model = JumpStartModel(model_id=model_id,model_version="1.1.1")

In [17]:
embedding_predictor = text_embedding_model.deploy(instance_type="ml.g5.2xlarge")

---------!

In [18]:
import json
import sagemaker

from langchain_core.prompts import PromptTemplate
from langchain_community.llms import SagemakerEndpoint
from langchain_community.embeddings import SagemakerEndpointEmbeddings
from langchain_community.llms.sagemaker_endpoint import LLMContentHandler
from langchain_community.embeddings.sagemaker_endpoint import EmbeddingsContentHandler

In [19]:
sess = sagemaker.session.Session()
region = sess._region_name

In [20]:
llm_endpoint_name = "meta-textgeneration-llama-3-8b-instruct-2024-10-21-01-56-36-789"
embedding_endpoint_name = "hf-sentencesimilarity-bge-large-en-v1-5-2024-10-21-02-07-53-002"

In [21]:
from typing import Dict

class Llama38BContentHandler(LLMContentHandler):
    content_type = "application/json"
    accepts = "application/json"

    def transform_input(self, prompt: str, model_kwargs: dict) -> bytes:
        payload = {
            "inputs": prompt,
            # default parameters
            "parameters": {
                # Controls the maximum number of tokens the model can generate
                "max_new_tokens": 700, #started at 1000
                # lower value makes the model more deterministic / higher value allows for more diverse responses
                "top_p": 0.9,
                # randomness
                "temperature": 0.4, #started at 0.6
                "stop": ["<|eot_id|>"],
            },
        }
        input_str = json.dumps(
            payload,
        )
        print(input_str)
        return input_str.encode("utf-8")

    def transform_output(self, output: bytes) -> str:
        response_json = json.loads(output.read().decode("utf-8"))
        print(response_json)
        content = response_json["generated_text"].strip()
        return content

In [22]:
# Instantiate the content handler for Llama3-8B
llama_content_handler = Llama38BContentHandler()

# Setup for using the Llama3-8B model with SageMaker Endpoint
llm = SagemakerEndpoint(
     endpoint_name=llm_endpoint_name,
     region_name=region,
     model_kwargs={"max_new_tokens": 1024, "top_p": 0.9, "temperature": 0.4},
     content_handler=llama_content_handler
 )

In [23]:
from typing import List

class BGEContentHandlerV15(EmbeddingsContentHandler):
    content_type = "application/json"
    accepts = "application/json"

    def transform_input(self, text_inputs: List[str], model_kwargs: dict) -> bytes:
        """
        Transforms the input into bytes that can be consumed by SageMaker endpoint.
        Args:
            text_inputs (list[str]): A list of input text strings to be processed.
            model_kwargs (Dict): Additional keyword arguments to be passed to the endpoint.
               Possible keys and their descriptions:
               - mode (str): Inference method. Valid modes are 'embedding', 'nn_corpus', and 'nn_train_data'.
               - corpus (str): Corpus for Nearest Neighbor. Required when mode is 'nn_corpus'.
               - top_k (int): Top K for Nearest Neighbor. Required when mode is 'nn_corpus'.
               - queries (list[str]): Queries for Nearest Neighbor. Required when mode is 'nn_corpus' or 'nn_train_data'.
        Returns:
            The transformed bytes input.
        """
        input_str = json.dumps(
            {
                "text_inputs": text_inputs,
                **model_kwargs
            }
        )
        return input_str.encode("utf-8")

    def transform_output(self, output: bytes) -> List[List[float]]:
        """
        Transforms the bytes output from the endpoint into a list of embeddings.
        Args:
            output: The bytes output from SageMaker endpoint.
        Returns:
            The transformed output - list of embeddings
        Note:
            The length of the outer list is the number of input strings.
            The length of the inner lists is the embedding dimension.
        """
        response_json = json.loads(output.read().decode("utf-8"))
        return response_json["embedding"]

In [24]:
bge_content_handler = BGEContentHandlerV15()
sagemaker_embeddings = SagemakerEndpointEmbeddings(
    endpoint_name=embedding_endpoint_name,
    region_name=region,
    model_kwargs={"mode": "embedding"},
    content_handler=bge_content_handler,
)

In [25]:
!pip install openpyxl

  Using cached openpyxl-3.1.5-py2.py3-none-any.whl.metadata (2.5 kB)
  Using cached et_xmlfile-1.1.0-py3-none-any.whl.metadata (1.8 kB)
Using cached openpyxl-3.1.5-py2.py3-none-any.whl (250 kB)
Using cached et_xmlfile-1.1.0-py3-none-any.whl (4.7 kB)


In [26]:
import os

# Set the directory path to the correct folder
#directory_path = 'user-default-efs'

contents = os.listdir()
# Filter the list to get only PDF files
xlsx_files = [item for item in contents if item.endswith('.xlsx')]

# Print the PDF files
print("Excel files in the specified directory:")
for item in xlsx_files:
    print(item)

Excel files in the specified directory:
recipe_final1_split_1.xlsx
recipe_final1_split_12.xlsx
recipe_final1_split_4.xlsx
recipe_final1_split_6.xlsx
recipe_final2_split_1.xlsx
recipe_final1_split_8.xlsx
recipe_final1_split_7.xlsx
recipe_final1_split_5.xlsx
recipe_final1_split_9.xlsx
recipe_final1_split_10.xlsx
recipe_final2_split_2.xlsx
recipe_final1_split_3.xlsx
recipe_final2_split_3.xlsx
recipe_final1_split_11.xlsx
recipe_final2_split_5.xlsx
recipe_final2_split_6.xlsx
recipe_final2_split_4.xlsx
recipe_final2_split_10.xlsx
recipe_final2_split_7.xlsx
recipe_final2_split_9.xlsx
recipe_final2_split_11.xlsx
recipe_final2_split_14.xlsx
recipe_final2_split_12.xlsx
recipe_final1_split_2.xlsx
recipe_final2_split_16.xlsx
recipe_final2_split_15.xlsx
recipe_final2_split_13.xlsx
recipe_final2_split_8.xlsx


In [27]:
import json
from typing import List, Dict
from langchain import PromptTemplate
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain_community.vectorstores import FAISS
from langchain.indexes.vectorstore import VectorStoreIndexWrapper
from langchain.schema import HumanMessage
import pandas as pd
from langchain.docstore.document import Document

# Define the optimized prompt template
prompt_template = """<|begin_of_text|><|start_header_id|>system<|end_header_id|>
You are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:
1. Recipe Name
2. Ingredients (as a numbered list)
3. Instructions (as a numbered list)
4. Any additional notes or tips
Ensure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.
Use the following context to answer the user's query:
{context}
<|eot_id|><|start_header_id|>user<|end_header_id|>
{query}
<|eot_id|><|start_header_id|>assistant<|end_header_id|>
Provide a complete recipe following the structure outlined above.
"""

PROMPT = PromptTemplate(
    template=prompt_template, input_variables=["context", "query"]
)

# 1. Document Processing
documents = []
metadata = [{} for _ in xlsx_files]  # Initialize metadata for each Excel file

for idx, file in enumerate(xlsx_files):
    # Load Excel file
    df = pd.read_excel(file)
    
    # Process each row in the Excel file
    for _, row in df.iterrows():
        recipe = f"Title: {row['title']}\nIngredients: {row['ingredients']}\nInstructions: {row['instructions']}"
        document = Document(page_content=recipe, metadata=metadata[idx])
        documents.append(document)

# Apply text splitting if needed
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1000,
    chunk_overlap=50,
)
docs = text_splitter.split_documents(documents)

# 2. Vector Store Creation
vectorstore_faiss = FAISS.from_documents(
    docs,
    sagemaker_embeddings,
)
wrapper_store_faiss = VectorStoreIndexWrapper(vectorstore=vectorstore_faiss)

# 3. Load Allergens
with open('allergens.json', 'r') as f:
    allergens = json.load(f)

# 4. Define Filtering and Query Functions
def filter_documents_by_allergen(docs, allergen):
    if allergen:
        return [doc for doc in docs if allergen.lower() not in doc.page_content.lower()]
    return docs

def query_with_allergen_filter(query: str, allergen: str = None):
    # Retrieve relevant documents
    retrieved_docs = wrapper_store_faiss.vectorstore.similarity_search(query)
    
    # Filter documents based on allergen
    filtered_docs = filter_documents_by_allergen(retrieved_docs, allergen)
    
    # Generate response using filtered documents
    context = "\n".join([doc.page_content for doc in filtered_docs])
    
    formatted_prompt = PROMPT.format(context=context, query=query)
    
    # Use the invoke method instead of calling the LLM directly
    response = llm.invoke([HumanMessage(content=formatted_prompt)])
    
    # Check if response is a string or an object with content attribute
    if isinstance(response, str):
        return response
    elif hasattr(response, 'content'):
        return response.content
    else:
        return str(response)  # fallback to string representation

# Modify the process_query function to include the allergen information in the query
def process_query(user_query: str):
    allergen = None
    if "allergic to" in user_query.lower():
        allergen = user_query.lower().split("allergic to")[-1].split(",")[0].strip()
        query = f"Please provide a recipe that doesn't include {allergen}. {user_query}"
    else:
        query = user_query
    
    result = query_with_allergen_filter(query, allergen)
    return result

In [29]:
# Example Usage
query = "I am allergic to milk, please give me a dairy-free recipe."
result = process_query(query)
print(result)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nTitle: No-Bake Cheesecake Squares\nIngredients: 1 stick plus 6 tablespoons unsalted butter, at room temperature, 1/4 cup granulated sugar, 1/4 cup unsweetened cocoa powder, 1 large egg, lightly beaten, 2 cups graham cracker crumbs, 3/4 cup sweetened shredded coconut, 1/2 cup walnuts, finely chopped, plus whole nuts for topping, 4 ounces cream cheese, at room temperature, 1/2 teaspoon vanilla extract, Pinch of salt, 1 1/2 cups confectioners' sugar, 1 tablespoon heavy cream, 4 ounce

In [30]:
# Another Example Usage run previously by Phoebe
query = "I have brussel sprouts and carrots, can you give me a healthy dairy-free recipe to make."
result = process_query(query)
print(result)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nTitle: Sprouts, Squash and 'Shrooms\nIngredients: 1 pint Brussels sprouts, tough outer leaves removed and halved from top to bottom, 1 pound baby carrots, 1 tablespoon olive oil, 1 tablespoon unsalted butter, 2 shallots, sliced, 2 cloves garlic, gently crushed with the side of a knife and minced, 1 pint (about 6 ounces) white mushrooms, brushed clean, trimmed, and halved if medium or sliced, if large (leave whole if small), 1 cup white wine, 5 or 6 small pattypan squash, quartered

In [170]:
# Example with multiple different queries
queries = [
    "Give me a recipe for a vegetarian pasta dish.",
    "I'm allergic to nuts. Can you give me a dessert recipe?",
    "I need a gluten-free breakfast recipe."
]

for q in queries:
    print(f"\nQuery: {q}")
    result = process_query(q)
    print(result)
    print("-" * 50)


Query: Give me a recipe for a vegetarian pasta dish.
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nTitle: Baked Pasta with Tomatoes, Shiitakes, and Prosciutto\nIngredients: 2 cups finely chopped onion, 2 large garlic cloves, minced, 1/4 teaspoon dried hot red pepper flakes, or to taste, 1 teaspoon dried basil, crumbled, 1 teaspoon dried oregano, crumbled, 2 tablespoons olive oil, 1 pound fresh shiitakes, stems discarded and the caps sliced, 8 ounces (1 stick) unsalted butter, 6 tablespoons all-purpose flour, 4 cups milk, 2 (

In [33]:
# Example with multiple equal queries (same request)
queries = [
    "I'm allergic to nuts. Can you give me a dessert recipe?",
    "I'm allergic to nuts. Can you give me a dessert recipe?",
    "I'm allergic to nuts. Can you give me a dessert recipe?"
]

for q in queries:
    print(f"\nQuery: {q}")
    result = process_query(q)
    print(result)
    print("-" * 50)


Query: I'm allergic to nuts. Can you give me a dessert recipe?
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nTitle: Honey-Apple-Almond Cake\nIngredients: 4 large egg whites, 1 cup sugar, 1 teaspoon almond extract, 3 sticks unsalted butter, cut into small pieces, at room temperature, 3 tablespoons honey, Basic Vanilla Cake, recipe follows, baked and cooled, 3/4 cup apple butter, Sliced almonds, toasted, for topping, 2 sticks unsalted butter, at room temperature, plus more for the pans, 3 cups all-purpose flour, plus more for 

In [31]:
# Another Example Usage run previously by Phoebe
query = "How to cook a dish with pork tenderloin?"
result = process_query(query)
print(result)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nInstructions: In a deep rectangular or oval dish that can contain the meat and all the other ingredients except for the mushrooms, combine the pork tenderloin, 3 tablespoons of olive oil, onion, garlic, bay leaves, celery, rosemary, and red wine. Turn the meat over several times to coat it well, then cover the dish with plastic wrap and refrigerate overnight. Take it out occasionally whenever convenient to turn the pork over, basting it with its marinade.\nThe following day, take 

In [173]:
# Another query example
query = "I have brussel sprouts and carrots, can you give me a healthy dairy-free recipe to make."
result = process_query(query)
print(result)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nInstructions: Bring a stockpot of lightly salted water to a boil over high heat. Add the potatoes, and cook until tender, 15 to 20 minutes. Drain potatoes. Cut potatoes into 1/3-inch-thick rounds, and transfer to a bowl. Set aside. Bring a stockpot of lightly salted water to a boil over high heat, and prepare an ice bath. Trim bottoms from Brussels sprouts. Peel away the inner leaves, one by one. Add carrots to boiling water, and cook until tender, about 4 minutes. Using a slotted

In [32]:
# Another Example Usage run previously by Phoebe
query = "How to cook a dish with pork tenderloin?"
result = process_query(query)
print(result)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nInstructions: In a deep rectangular or oval dish that can contain the meat and all the other ingredients except for the mushrooms, combine the pork tenderloin, 3 tablespoons of olive oil, onion, garlic, bay leaves, celery, rosemary, and red wine. Turn the meat over several times to coat it well, then cover the dish with plastic wrap and refrigerate overnight. Take it out occasionally whenever convenient to turn the pork over, basting it with its marinade.\nThe following day, take 

## Metric

* Blue

In [35]:
import nltk
nltk.download('punkt')

[nltk_data] Downloading package punkt to /home/sagemaker-
[nltk_data]     user/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


True

In [45]:
import random
import re
from nltk.translate.bleu_score import sentence_bleu

def simple_tokenize(text):
    return re.findall(r'\w+|[^\w\s]', text.lower())

def calculate_bleu(reference, candidate):
    reference_tokens = [simple_tokenize(reference)]
    candidate_tokens = simple_tokenize(candidate)
    return sentence_bleu(reference_tokens, candidate_tokens)

def evaluate_recipes_bleu(reference_recipes, num_recipes=10):
    # Sample queries
    queries = []
    for recipe in reference_recipes:
        title = recipe.split('\n')[0].split(': ')[1]
        queries.append(f"Give me a recipe for {title}")
    
    sampled_queries = random.sample(queries, min(num_recipes, len(queries)))
    
    bleu_scores = []
    for i, query in enumerate(sampled_queries, 1):
        generated_recipe = process_query(query)  # RAG system's query processing function
        
        # Find the most similar reference recipe
        most_similar_recipe = max(reference_recipes, 
                                  key=lambda ref: calculate_bleu(ref, generated_recipe))
        
        bleu_score = calculate_bleu(most_similar_recipe, generated_recipe)
        bleu_scores.append(bleu_score)
        
        print(f"\nRecipe {i}:")
        print(f"Query: {query}")
        print(f"Most Similar Reference Recipe (truncated): {most_similar_recipe[:200]}...")
        print(f"Generated Recipe (truncated): {generated_recipe[:200]}...")
        print(f"BLEU score: {bleu_score}")
        print("-" * 50)
    
    avg_bleu = sum(bleu_scores) / len(bleu_scores)
    print(f"\nAverage BLEU score: {avg_bleu}")
    return bleu_scores, avg_bleu

# Run the evaluation
bleu_scores, avg_bleu = evaluate_recipes_bleu(reference_recipes, num_recipes=10)

{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instructions are always presented as numbered lists for clarity and ease of use.\nUse the following context to answer the user's query:\nTitle: Spatchcocked Turkey with Sage Butter and Gravy\nIngredients: One 14-pound turkey, Sage Butter, recipe follows, Olive oil, Salt and freshly ground black pepper, 5 bay leaves, 4 cloves garlic, smashed, 2 Granny Smith apples, cut into 1/2-inch dice, 2 ribs celery, cut into 1/2-inch dice, 1 large or 2 small carrots, peeled and cut into 1/2-inch dice, 1 large onion, cut into 1/2-inch dice, 1 cinnamon stick, 1 bunch fresh thyme, Kosher salt, 4 cups chicken stock, 2 cups apple cid

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 1:
Query: Give me a recipe for Spatchcocked Turkey with Sage Butter and Gravy
Most Similar Reference Recipe (truncated): Title: Spatchcocked Turkey with Sage Butter and Gravy
Ingredients: One 14-pound turkey, Sage Butter, recipe follows, Olive oil, Salt and freshly ground black pepper, 5 bay leaves, 4 cloves garlic, sma...
Generated Recipe (truncated): **Spatchcocked Turkey with Sage Butter and Gravy**

**Ingredients:**

1. One 14-pound turkey
2. Sage Butter (recipe follows)
3. Olive oil
4. Salt and freshly ground black pepper
5. 5 bay leaves
6. 4 c...
BLEU score: 0.26848698745971405
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\n

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 2:
Query: Give me a recipe for Corn Pudding
Most Similar Reference Recipe (truncated): Title: Corn Pudding
Ingredients: 3 eggs, 1 cup heavy cream, 2 tablespoons sugar, 1 1/2 tablespoons flour, 1 teaspoon salt, 1 teaspoon baking powder, 3 cups corn, 1/4 pound melted butter
Instructions: ...
Generated Recipe (truncated): **Corn Pudding**

**Ingredients:**

1. 3 eggs
2. 1 cup heavy cream
3. 2 tablespoons sugar
4. 1 1/2 tablespoons flour
5. 1 teaspoon salt
6. 1 teaspoon baking powder
7. 3 cups corn
8. 1/4 pound melted b...
BLEU score: 0.27569386939150375
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and instru

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 3:
Query: Give me a recipe for Marble Cheesecake
Most Similar Reference Recipe (truncated): Title: Marble Cheesecake
Ingredients: 1 1/2 cups graham cracker crumbs, 6 tablespoon unsalted butter, melted, 1/4 cup sugar, 2 squares unsweetened chocolate, 2 tablespoon unsalted butter, 3 8-ounce pa...
Generated Recipe (truncated): **Marble Cheesecake**

**Ingredients:**

1. 1 1/2 cups graham cracker crumbs
2. 6 tablespoons unsalted butter, melted
3. 1/4 cup sugar
4. 2 squares unsweetened chocolate
5. 2 tablespoons unsalted butt...
BLEU score: 0.36451887750067735
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and i

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 4:
Query: Give me a recipe for New Zealand Venison Sliders with Tomato Chile Jam
Most Similar Reference Recipe (truncated): Title: New Zealand Venison Sliders with Tomato Chile Jam
Ingredients: About 1 pound (500g) ground beef, About 1 pound (500g) farmed ground New Zealand venison (any cut will do but the most cost effect...
Generated Recipe (truncated): **New Zealand Venison Sliders with Tomato Chile Jam**

**Ingredients:**

1. About 1 pound (500g) ground beef
2. About 1 pound (500g) farmed ground New Zealand venison (any cut will do but the most cos...
BLEU score: 0.3259071379123106
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 5:
Query: Give me a recipe for Jalapeno Pesto
Most Similar Reference Recipe (truncated): Title: Jalapeno Pesto
Ingredients: 1/4 cup walnuts , 2 cloves garlic , 2 cups packed fresh basil leaves , 3/4 cup shredded Parmagiano-Reggiano cheese , 1 jalapeno pepper, stem removed , 2/3 cup olive ...
Generated Recipe (truncated): **Jalapeno Pesto**

**Ingredients:**

1. 1/4 cup walnuts
2. 2 cloves garlic
3. 2 cups packed fresh basil leaves
4. 3/4 cup shredded Parmesan cheese
5. 1 jalapeno pepper, stem removed
6. 2/3 cup olive ...
BLEU score: 0.32115936031893466
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and inst

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 6:
Query: Give me a recipe for Banana Pancakes II
Most Similar Reference Recipe (truncated): Title: Banana Pancakes
Ingredients: 1 1/2 cups all-purpose flour, 1 tablespoon sugar, 1/2 teaspoon salt, 1 teaspoon baking soda, 1 teaspoon baking powder, 1 large egg, 1 cup buttermilk, 1/4 cup whole ...
Generated Recipe (truncated): **Banana Pancakes II**

**Ingredients:**

1. 1 1/2 cups all-purpose flour
2. 1 tablespoon sugar
3. 1/2 teaspoon salt
4. 1 teaspoon baking soda
5. 1 teaspoon baking powder
6. 1 large egg
7. 1 cup butte...
BLEU score: 0.22998276049191493
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and 

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 7:
Query: Give me a recipe for Eggnog Rum Muffins
Most Similar Reference Recipe (truncated): Title: Eggnog Rum Muffins
Ingredients: 2 cups all-purpose flour , 1/2 cup white sugar , 1 tablespoon baking powder , 1 teaspoon nutmeg , 1/2 teaspoon salt , 1/2 cup walnuts, chopped (optional) , 1 cup...
Generated Recipe (truncated): **Eggnog Rum Muffins**

**Ingredients:**

1. 2 cups all-purpose flour
2. 1/2 cup white sugar
3. 1 tablespoon baking powder
4. 1 teaspoon nutmeg
5. 1/2 teaspoon salt
6. 1/2 cup walnuts, chopped (option...
BLEU score: 0.49044436523127866
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredients and 

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 8:
Query: Give me a recipe for Peanut Meringues with Hot Chocolate
Most Similar Reference Recipe (truncated): Title: Peanut Meringues with Hot Chocolate: Cacahuetes Merengues con Chocolate Caliente
Ingredients: 1 cup egg whites, 1/4 teaspoon cream of tartar, 3/4 cup granulated sugar, 8 ounces powdered sugar, ...
Generated Recipe (truncated): **Peanut Meringues with Hot Chocolate: Cacahuetes Merengues con Chocolate Caliente**

**Ingredients:**

1. 1 cup egg whites
2. 1/4 teaspoon cream of tartar
3. 3/4 cup granulated sugar
4. 8 ounces powd...
BLEU score: 0.346323376088845
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that i

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 9:
Query: Give me a recipe for Cream Cheese Coffee Cake I
Most Similar Reference Recipe (truncated): Title: Cream Cheese Coffee Cake I
Ingredients: 1/3 cup packed dark brown sugar , 2 teaspoons unsweetened cocoa powder , 1/3 cup chopped semisweet chocolate , 1 teaspoon ground cinnamon , 1/3 cup raisi...
Generated Recipe (truncated): **Cream Cheese Coffee Cake I**

**Ingredients:**

1. 1/3 cup packed dark brown sugar
2. 2 teaspoons unsweetened cocoa powder
3. 1/3 cup chopped semisweet chocolate
4. 1 teaspoon ground cinnamon
5. 1/3...
BLEU score: 0.3643128082132172
--------------------------------------------------
{"inputs": "Human: <|begin_of_text|><|start_header_id|>system<|end_header_id|>\nYou are a helpful assistant specialized in providing recipes. When answering questions about recipes, always structure your response as follows:\n1. Recipe Name\n2. Ingredients (as a numbered list)\n3. Instructions (as a numbered list)\n4. Any additional notes or tips\nEnsure that ingredien

/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 4-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)
/opt/conda/lib/python3.11/site-packages/nltk/translate/bleu_score.py:577: UserWarning: 
The hypothesis contains 0 counts of 3-gram overlaps.
Therefore the BLEU score evaluates to 0, independently of
how many N-gram overlaps of lower order it contains.
Consider using lower n-gram order or use SmoothingFunction()
  warnings.warn(_msg)



Recipe 10:
Query: Give me a recipe for Turkey Meatballs
Most Similar Reference Recipe (truncated): Title: Classic Italian Turkey Meatballs
Ingredients: 1 cup plain breadcrumbs, 1/2 cup finely grated Parmesan, plus more for serving, 1/4 cup chopped fresh basil, 1/4 cup chopped fresh Italian parsley,...
Generated Recipe (truncated): **Turkey Meatballs**

**Ingredients:**

1. 1 green onion, thinly sliced
2. 1 minced clove of garlic
3. 1 egg, lightly beaten
4. 1-ounce fresh bread crumbs
5. 2 tablespoons grated Parmesan cheese, plus...
BLEU score: 0.26195883998105673
--------------------------------------------------

Average BLEU score: 0.32487883825894526


I think a BLUE csore of 0.325 is quite respectful.
a) Variability in recipes: There are many ways to express the same cooking instructions, and ingredient lists can vary in order and specificity.
b) Creative task: Recipe generation is more creative than straight translation, which BLEU was originally designed for.
c) Comparison to other NLP tasks: In many NLP tasks, scores around 0.3-0.4 are considered good, especially for creative or variable content generation.

This score suggests that your RAG system is capturing a significant amount of relevant information and structure from the reference recipes.

In [ ]:
# Calculate Avg. BLUE for 100 random recipes

#import random
#import re
#from nltk.translate.bleu_score import sentence_bleu
#from tqdm import tqdm  # for progress bar

#def simple_tokenize(text):
    #return re.findall(r'\w+|[^\w\s]', text.lower())

#def calculate_bleu(reference, candidate):
    #reference_tokens = [simple_tokenize(reference)]
    #candidate_tokens = simple_tokenize(candidate)
    #return sentence_bleu(reference_tokens, candidate_tokens)

#def evaluate_recipes_bleu(reference_recipes, num_recipes=100):
    # Sample queries
    #queries = [f"Give me a recipe for {recipe.split('\\n')[0].split(': ')[1]}" 
               #for recipe in reference_recipes]
    
    #sampled_queries = random.sample(queries, min(num_recipes, len(queries)))
    
    #bleu_scores = []
    #for query in tqdm(sampled_queries, desc="Evaluating Recipes"):
        #generated_recipe = process_query(query)  # Your RAG system's query processing function
        
        # Find the most similar reference recipe
        #most_similar_recipe = max(reference_recipes, 
                                  #key=lambda ref: calculate_bleu(ref, generated_recipe))
        
        #bleu_score = calculate_bleu(most_similar_recipe, generated_recipe)
        #bleu_scores.append(bleu_score)
    
    #avg_bleu = sum(bleu_scores) / len(bleu_scores)
    #return avg_bleu

# Run the evaluation
#avg_bleu = evaluate_recipes_bleu(reference_recipes, num_recipes=100)
#print(f"\nAverage BLEU score for 100 recipes: {avg_bleu}")


* ROUGE

* F1

* Embedding Quality:

Cosine similarity: Compare the embeddings of similar recipes to ensure they are close in the vector space

* Allergen Compliance

Allergen avoidance rate: Calculate the percentage of recipes that successfully avoid specified allergens when requested.

* Response Diversity:

Unique recipe rate: Measure the percentage of unique recipes generated across multiple similar queries to ensure variety.




* Retrieval Efficiency:

Retrieval time: Measure the time taken to retrieve relevant documents from the FAISS index.

* End-to-End Performance:

Query-to-recipe time: Measure the total time from receiving a query to generating a complete recipe